File: compat/suites/dotnet-sdk/Groups/StsGroup.cs
using Amazon.SecurityToken.Model;
using OvercastCompat.Clients;
using OvercastCompat.Harness;

namespace OvercastCompat.Groups;

public sealed class StsGroup(AwsClients clients) : IServiceGroup
{
    public IReadOnlyDictionary<string, TestFn> Impls() => new Dictionary<string, TestFn>(StringComparer.Ordinal)
    {
        ["GetCallerIdentity"] = GetCallerIdentityAsync,
        ["GetSessionToken"] = GetSessionTokenAsync,
        ["GetFederationToken"] = GetFederationTokenAsync,
        ["AssumeRole"] = AssumeRoleAsync,
        ["AssumeRoleWithWebIdentity"] = AssumeRoleWithWebIdentityAsync,
    };

    public IReadOnlyDictionary<string, SetupFn> Setups() => new Dictionary<string, SetupFn>(StringComparer.Ordinal);

    public IReadOnlyDictionary<string, SetupFn> Teardowns() => new Dictionary<string, SetupFn>(StringComparer.Ordinal);

    private async Task GetCallerIdentityAsync(TestContext context)
    {
        var response = await clients.STS().GetCallerIdentityAsync(new GetCallerIdentityRequest());
        Assertions.NotBlank(response.Account, "GetCallerIdentity: account");
        Assertions.NotBlank(response.Arn, "GetCallerIdentity: arn");
        Assertions.NotBlank(response.UserId, "GetCallerIdentity: userId");
    }

    private async Task GetSessionTokenAsync(TestContext context)
    {
        var response = await clients.STS().GetSessionTokenAsync(new GetSessionTokenRequest());
        Assertions.NotNull(response.Credentials, "GetSessionToken: credentials");
        Assertions.NotBlank(response.Credentials.AccessKeyId, "GetSessionToken: accessKeyId");
    }

    private async Task GetFederationTokenAsync(TestContext context)
    {
        var response = await clients.STS().GetFederationTokenAsync(new GetFederationTokenRequest
        {
            Name = "compat-user",
        });
        Assertions.NotNull(response.Credentials, "GetFederationToken: credentials");
        Assertions.NotBlank(response.Credentials.AccessKeyId, "GetFederationToken: accessKeyId");
    }

    private async Task AssumeRoleAsync(TestContext context)
    {
        var response = await clients.STS().AssumeRoleAsync(new AssumeRoleRequest
        {
            RoleArn = "arn:aws:iam::000000000000:role/compat-role",
            RoleSessionName = "dotnet-sdk-compat",
        });
        Assertions.NotNull(response.Credentials, "AssumeRole: credentials");
        Assertions.NotBlank(response.Credentials.AccessKeyId, "AssumeRole: accessKeyId");
    }

    private async Task AssumeRoleWithWebIdentityAsync(TestContext context)
    {
        var response = await clients.STS().AssumeRoleWithWebIdentityAsync(new AssumeRoleWithWebIdentityRequest
        {
            RoleArn = "arn:aws:iam::000000000000:role/compat-role",
            RoleSessionName = "dotnet-sdk-compat",
            WebIdentityToken = "fake-web-identity-token",
        });
        Assertions.NotNull(response.Credentials, "AssumeRoleWithWebIdentity: credentials");
        Assertions.NotBlank(response.Credentials.AccessKeyId, "AssumeRoleWithWebIdentity: accessKeyId");
    }
}

File: compat/suites/java-sdk/src/main/java/io/overcast/compat/groups/StsGroup.java
package io.overcast.compat.groups;

import io.overcast.compat.clients.AwsClients;
import io.overcast.compat.harness.Assertions;
import io.overcast.compat.harness.TestContext;
import io.overcast.compat.harness.TestFn;
import software.amazon.awssdk.services.sts.StsClient;

import java.util.Map;

/**
 * STS compatibility test group.
 *
 * <p>Groups: sts-identity, sts-assume.
 */
public final class StsGroup implements ServiceGroup {

    private final AwsClients clients;

    public StsGroup(AwsClients clients) {
        this.clients = clients;
    }

    private StsClient sts() { return clients.sts(); }

    @Override
    public Map<String, TestFn> impls() {
        return Map.ofEntries(
                Map.entry("GetCallerIdentity",        this::getCallerIdentity),
                Map.entry("GetSessionToken",          this::getSessionToken),
                Map.entry("GetFederationToken",       this::getFederationToken),
                Map.entry("AssumeRole",               this::assumeRole),
                Map.entry("AssumeRoleWithWebIdentity",this::assumeRoleWithWebIdentity)
        );
    }

    @Override
    public Map<String, TestFn> setups() {
        return Map.of();
    }

    @Override
    public Map<String, TestFn> teardowns() {
        return Map.of();
    }

    // ── sts-identity ──────────────────────────────────────────────────────────

    private void getCallerIdentity(TestContext ctx) throws Exception {
        var resp = sts().getCallerIdentity();
        Assertions.assertNotBlank(resp.account(), "GetCallerIdentity: account");
        Assertions.assertNotBlank(resp.arn(), "GetCallerIdentity: arn");
        Assertions.assertNotBlank(resp.userId(), "GetCallerIdentity: userId");
    }

    private void getSessionToken(TestContext ctx) throws Exception {
        var resp = sts().getSessionToken();
        Assertions.assertNotNull(resp.credentials(), "GetSessionToken: credentials");
        Assertions.assertNotBlank(resp.credentials().accessKeyId(), "GetSessionToken: accessKeyId");
    }

    private void getFederationToken(TestContext ctx) throws Exception {
        var resp = sts().getFederationToken(r -> r.name("compat-user"));
        Assertions.assertNotNull(resp.credentials(), "GetFederationToken: credentials");
        Assertions.assertNotBlank(resp.credentials().accessKeyId(), "GetFederationToken: accessKeyId");
    }

    // ── sts-assume ────────────────────────────────────────────────────────────

    private void assumeRole(TestContext ctx) throws Exception {
        var resp = sts().assumeRole(r -> r
                .roleArn("arn:aws:iam::000000000000:role/compat-role")
                .roleSessionName("java-sdk-compat"));
        Assertions.assertNotNull(resp.credentials(), "AssumeRole: credentials");
        Assertions.assertNotBlank(resp.credentials().accessKeyId(), "AssumeRole: accessKeyId");
    }

    private void assumeRoleWithWebIdentity(TestContext ctx) throws Exception {
        var resp = sts().assumeRoleWithWebIdentity(r -> r
                .roleArn("arn:aws:iam::000000000000:role/compat-role")
                .roleSessionName("java-sdk-compat")
                .webIdentityToken("fake-web-identity-token"));
        Assertions.assertNotNull(resp.credentials(), "AssumeRoleWithWebIdentity: credentials");
    }
}
